github/VisualStudio
# Allow user to specify default remote when there's no remote named "origin"

### What this PR does

Some users like to rename their "origin" remote so they're forced to be explicit about which repository they're pushing to. When this happend the extension wasn't able to detect their repository originated from GitHub.

This PR takes advantage of the fact the order of remotes doesn't change when a remote is renamed, added or deleted. When a user renames their "origin" remote, it will remain the first defined remote. We can fall back to using the first remote as the originating remote/clone URL.

- Use the first remote as the originating remote when no "origin" is defined

### TODO

The default remote "origin" is hard-coded in a bunch of other places:

```
  C:\Source\github.com\github\VisualStudio\src\GitHub.App\Services\PullRequestService.cs(290):                        var remote = await gitClient.GetHttpRemote(repo, "origin");
  C:\Source\github.com\github\VisualStudio\src\GitHub.App\Services\PullRequestService.cs(374):                    var remote = await gitClient.GetHttpRemote(repo, "origin");
  C:\Source\github.com\github\VisualStudio\src\GitHub.App\Services\PullRequestService.cs(440):                        var remote = await gitClient.GetHttpRemote(repo, "origin");
  C:\Source\github.com\github\VisualStudio\src\GitHub.App\Services\PullRequestService.cs(500):                    var remote = await gitClient.GetHttpRemote(repo, "origin");
  C:\Source\github.com\github\VisualStudio\src\GitHub.App\Services\PullRequestService.cs(613):                var remote = await gitClient.GetHttpRemote(repo, "origin");
  C:\Source\github.com\github\VisualStudio\src\GitHub.App\Services\PullRequestService.cs(666):                var remote = await gitClient.GetHttpRemote(repo, "origin");
  C:\Source\github.com\github\VisualStudio\src\GitHub.App\Services\RepositoryForkService.cs(79):                            var remote = await gitClient.GetHttpRemote(activeRepo, "origin");
  C:\Source\github.com\github\VisualStudio\src\GitHub.App\Services\RepositoryForkService.cs(106):            await gitClient.SetRemote(repository, "origin", originUri);
  C:\Source\github.com\github\VisualStudio\src\GitHub.App\Services\RepositoryPublishService.cs(50):                                     await gitClient.SetRemote(repo.LocalRepo, "origin", new Uri(repo.RemoteRepo.CloneUrl));
  C:\Source\github.com\github\VisualStudio\src\GitHub.App\Services\RepositoryPublishService.cs(51):                                     await gitClient.Push(repo.LocalRepo, "master", "origin");
  C:\Source\github.com\github\VisualStudio\src\GitHub.App\Services\RepositoryPublishService.cs(52):                                     await gitClient.Fetch(repo.LocalRepo, "origin");
  C:\Source\github.com\github\VisualStudio\src\GitHub.App\Services\RepositoryPublishService.cs(53):                                     await gitClient.SetTrackingBranch(repo.LocalRepo, "master", "origin");
  C:\Source\github.com\github\VisualStudio\src\GitHub.InlineReviews\Services\PullRequestSessionService.cs(237):                    await gitClient.Fetch(repo, "origin", sha, pullHeadRef);
```

### How to test

1. `git remote rename origin renamed_origin`
2. Confirm that `GitHub` section doesn't appear on `Team Explorer - Home`
3. Install the build from this PR
4. Confirm that `GitHub` section now appears on `Team Explorer - Home`

Fixes #1729
